pkg/provider/identity.go

```diff
@@ -35,26 +54,235 @@ type PeerIdentity struct {
+	Difficulty uint16 `default:"24" help:"minimum difficulty for identity generation"`
+	Version    string `help:"semantic version of CA storage format"`
 }
 
 // IdentityConfig allows you to run a set of Responsibilities with the given
 // identity. You can also just load an Identity from disk.
 type IdentityConfig struct {
 	CertPath string `help:"path to the certificate chain for this identity" default:"$CONFDIR/identity.leaf.cert"`
 	KeyPath  string `help:"path to the private key for this identity" default:"$CONFDIR/identity.leaf.key"`
+	Version  string `help:"semantic version of identity storage format"`
 	Address  string `help:"address to listen on" default:":7777"`
 }
 
-// LoadIdentity loads a FullIdentity from the given configuration
-func (ic IdentityConfig) LoadIdentity() (*FullIdentity, error) {
-	pi, err := FullIdentityFromFiles(ic.CertPath, ic.KeyPath)
+// GenerateCA creates a new full identity with the given difficulty
+func GenerateCA(ctx context.Context, difficulty uint16, concurrency uint) *CertificateAuthority {
+	if ctx == nil {
+		ctx = context.Background()
+	}
+	ctx, cancel := context.WithCancel(ctx)
+
+	caC := make(chan CertificateAuthority, 1)
+	for i := 0; i < int(concurrency); i++ {
+		go generateCAWorker(ctx, difficulty, caC)
+	}
+
+	ca := <-caC
+	cancel()
+
+	return &ca
+}
+
+// FullIdentityFromPEM loads a FullIdentity from a certificate chain and
+// private key file
+func FullIdentityFromPEM(chainPEM, keyPEM []byte) (*FullIdentity, error) {
+	cb, err := decodePEM(chainPEM)
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+	kb, err := decodePEM(keyPEM)
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+
+	// NB: there shouldn't be multiple keys in the key file but if there
+	// are, this uses the first one
+	pk, err := x509.ParseECPrivateKey(kb[0])
+	if err != nil {
+		return nil, errs.New("unable to parse EC private key", err)
+	}
+
+	pi, err := PeerIdentityFromCertChain(cb)
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+
+	return &FullIdentity{
+		PeerIdentity: *pi,
+		PrivateKey:   pk,
+	}, nil
+}
+
+// PeerIdentityFromCertChain loads a PeerIdentity from a chain of certificates
+func PeerIdentityFromCertChain(chain [][]byte) (*PeerIdentity, error) {
+	ca, err := x509.ParseCertificate(chain[1])
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+
+	l, err := x509.ParseCertificate(chain[0])
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+
+	return PeerIdentityFromCerts(l, ca)
+}
+
+// PeerIdentityFromCerts loads a PeerIdentity from a pair of leaf and ca x509 certificates
+func PeerIdentityFromCerts(leaf, ca *x509.Certificate) (*PeerIdentity, error) {
+	i, err := idFromCert(ca)
+	if err != nil {
+		return nil, err
+	}
+
+	return &PeerIdentity{
+		CA: CertificateAuthority{
+			Cert: ca,
+			ID:   i,
+		},
+		Leaf: leaf,
+	}, nil
+}
+
+// VerifyPeerIdentityFunc returns a function to use with `tls.Certificate.VerifyPeerCertificate`
+// that verifies the peer identity satisfies the minimum difficulty
```

Comment:
just want to reiterate that regardless of the minimum difficulty of a node you will want to store to, there will always be cases where you'll want to allow ssl communication with nodes that have difficulty 0, so in practice we can't ever use this.

Reply:
What about just passing a 0 difficulty?: https://sourcegraph.com/github.com/storj/storj@f5f7f15e9a2eba1939bab25be74cebc282ead81e/-/blob/pkg/provider/provider.go#L34

Reply:
why have it though? it will always be a zero difficulty